### PR TITLE
PHP8 Features

### PR DESCRIPTION
**Pull request by iBotPeaches**
* Since we dropped less than 8 - refactoring to support some newer language features.
* Also prevents hard-coding of User model via static property on `Logger`

### CODE REVIEW

**Review by justinvoelkel — APPROVED** [review]

File: src/Helpers/AuditResolver.php [hunk 1]
```diff
@@ -7,11 +7,7 @@
 
 class AuditResolver
 {
-    /**
-     * @param Model|null $model
-     * @return int|null
-     */
-    public static function findUserId($model): ?int
+    public static function findUserId(Model|null $model): ?int
```
Comment (charlesreffett) [inline src/Helpers/AuditResolver.php:10]: What about nullable type hint?
```suggestion
    public static function findUserId(?Model $model): ?int
```

File: src/Models/AuditActivity.php [hunk 2]
```diff
@@ -43,20 +46,20 @@ class AuditActivity extends BaseModel
...
+    protected $casts = [
+        'created_at' => 'datetime'
     ];
 
     //--------------------------------------------------------------------------------------------------------------
     // Mutators
     //--------------------------------------------------------------------------------------------------------------
 
-    protected function setIpAddressAttribute($value)
+    protected function setIpAddressAttribute(string|null $value): void
```
Comment (charlesreffett) [inline src/Models/AuditActivity.php:57]: Nullable type hint here too?
```suggestion
    protected function setIpAddressAttribute(?string $value): void
```

File: src/Models/AuditActivity.php [hunk 3]
```diff
@@ -43,20 +46,20 @@ class AuditActivity extends BaseModel
...
     {
-        $this->attributes['ip_address'] = inet_pton($value);
+        $this->attributes['ip_address'] = inet_pton((string)$value);
```
Comment (charlesreffett) [inline src/Models/AuditActivity.php:59]: Isn't a null value coerced to an empty string anyway?

File: src/Models/AuditActivity.php [hunk 4]
```diff
@@ -68,75 +71,49 @@ public function setUpdatedAtAttribute($value)
...
-        }
+        return match ($this->type) {
+            ActivityType::FAILED_LOGIN => trans('logger::enums.activity_type_failed_login'),
+            ActivityType::LOGOUT => trans('logger::enums.activity_type_logout'),
+            ActivityType::SUCCESSFUL_LOGIN => trans('logger::enums.activity_type_logged_in'),
+            ActivityType::LOCKED_OUT => trans('logger::enums.activity_type_locked_out'),
+            ActivityType::PASSWORD_CHANGE => trans('logger::enums.activity_type_password_change'),
+            ActivityType::GET_DATA => trans('logger::enums.activity_type_get_data'),
+            ActivityType::MODIFY_DATA => trans('logger::enums.activity_type_modify_data'),
+            default => throw new \Exception('Unknown enum type: ' . $this->type),
```
Comment (charlesreffett) [inline src/Models/AuditActivity.php:93]: Could we simplify the namespace?

File: src/Traits/HasRelationships.php [hunk 5]
```diff
@@ -8,13 +8,7 @@
...
 trait HasRelationships
 {
-    /**
-     * Retrieve the actual class name for a given morph class.
-     *
-     * @param  string  $class
-     * @return string
-     */
-    public static function getActualClassNameForMorph($class)
+    public static function getActualClassNameForMorph($class): string
```
Comment (charlesreffett) [inline src/Traits/HasRelationships.php:11]: What is `$class`  a string? Could we get a type hint?

Comment (iBotPeaches) [reply src/Traits/HasRelationships.php:11]: I can't on this one because it extends a parent function :/

File: src/Models/AuditActivity.php (hunk 2 above)
Comment (iBotPeaches) [reply src/Models/AuditActivity.php:57]: Is this just sugar? Or is anything different from the two? Just curious as `?|string` is not a valid docblock, but `string|null` is.

File: src/Models/AuditActivity.php (hunk 3 above)
Comment (iBotPeaches) [reply src/Models/AuditActivity.php:59]:
```
Deprecated: inet_pton(): Passing null to parameter #1 ($ip) of type string is deprecated in /home/user/scripts/code.php on line 2
```

Sadly not :(

File: src/Models/AuditActivity.php (hunk 2 above)
Comment (charlesreffett) [reply src/Models/AuditActivity.php:57]: It's syntactic sugar
> A single base type declaration can be marked nullable by prefixing the type with a question mark (?). Thus ?T and T|null are identical.

**Review by charlesreffett — APPROVED** [review]

File: src/Models/AuditActivity.php (hunk 2 above)
Comment (iBotPeaches) [reply src/Models/AuditActivity.php:57]: Okay, so it seems it works in userland, just docblock parsing doesn't work yet for those - https://github.com/laravel/framework/pull/45677/files

File: src/Models/AuditActivity.php (hunk 4 above)
Comment (iBotPeaches) [reply src/Models/AuditActivity.php:93]: I'll bring up a new PR to handle this and the nullable typehints

**Pull request merged**